Adding the Program.cs file to each module was the issue.

amirhossein18121380 opened the pull request:
Title: Added Extra API Layer for Better Organization

issue: https://github.com/meysamhadeli/booking-modular-monolith/issues/2

Description:
Added Program.cs to each module and implemented an extra API layer for better organization and maintainability. Also we could have achieved this with a single API layer, but this approach provides a clearer separation of concerns and easier integration with related libraries. I'm open to feedback and discussion on this change, and would appreciate any suggestions on how to improve the design.

meysamhadeli commented:
> Title: Added Extra API Layer for Better Organization
> 
> Description: Added Program.cs to each module and implemented an extra API layer for better organization and maintainability. Also we could have achieved this with a single API layer, but this approach provides a clearer separation of concerns and easier integration with related libraries. I'm open to feedback and discussion on this change, and would appreciate any suggestions on how to improve the design.

Hi, Thanks for contribution :)
My goal is having 1 API Layer. And each module have own separate Program.cs. and with this approach each module with using Program.cs can host itself and we don't need to have separate API layer for each module and we need 1 API layer for coordination all of the module and having multiple API leads us to having separated microservices not modular monolith.

amirhossein18121380 commented:
I deleted the extra API layer that I had added to each module before, and I added a Program.cs to each module (related class library). Now, each module can be hosted and executed independently.

However, I don't fully understand why we should add Program.cs to a class library (each module) if it cannot be hosted and run on its own. Wouldn't it be more efficient to keep an extra layer, such as a console application or API, that can be executed or hosted directly and reference the related class library? Am I right or wrong in my understanding?

meysamhadeli commented on src/Modules/Booking/src/Booking/Program.cs:
Please use top level statement and. Like Api layer.

meysamhadeli commented on src/Modules/Flight/src/Flight/Flight.csproj:
```diff
@@ -1,9 +1,22 @@
+	  <TargetFramework>net6.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
```
Please remove this item group, It's not necessary.

meysamhadeli commented on src/Modules/Flight/src/Flight/Flight.csproj:
```diff
@@ -1,9 +1,22 @@
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
+	  <OutputType>Exe</OutputType>
```
Please remove OutputType tag, it's not necessary.

amirhossein18121380 replied:
> Please remove OutputType tag, it's not necessary.

I applied the changes and I used top level statement like Api layer for program.cs but 
for top-level statements in C# to work, the project must be configured as an executable. By removing this tag, the project will default to a class library, which does not support top-level statements. 
The <OutputType> tag specifies whether the project is a library or an executable. 
With the <OutputType> set to Exe, our top-level statements in Program.cs will work as intended.

Should I remove the <OutputType> tag or leave it as is?

meysamhadeli replied:
Here in Api layer we don't have any additional configuration and it can host as well. also you need to add Properties folder and add configuration json file for host like Api layer. You need the follow the configuration exactly like Api layer and also add Properties folder and afterward each module should host properly:
[Api.csproj](https://github.com/meysamhadeli/booking-modular-monolith/blob/main/src/Api/src/Api.csproj)

amirhossein18121380 replied:
I followed your instructions and added a Properties folder with a configuration JSON file to each module, similar to the API layer. I also changed the project structure to Microsoft.NET.Sdk.Web. However, I'm not sure if everything is correct?

meysamhadeli replied:
> I followed your instructions and added a Properties folder with a configuration JSON file to each module, similar to the API layer. I also changed the project structure to Microsoft.NET.Sdk.Web. However, I'm not sure if everything is correct? for launchSetting i changed the port for each module. should all ports be exactly the same as api layer?

Thanks. Port should be different for each module. everything's seems good.

meysamhadeli commented:
> I deleted the extra API layer that I had added to each module before, and I added a Program.cs to each module (related class library). Now, each module can be hosted and executed independently.
> 
> However, I don't fully understand why we should add Program.cs to a class library (each module) if it cannot be hosted and run on its own. Wouldn't it be more efficient to keep an extra layer, such as a console application or API, that can be executed or hosted directly and reference the related class library? Am I right or wrong in my understanding?

If we have a separate Api for each module it provides complexity and we don't need it and we just need a simple Program.cs class for running independently each module. (it's also not very force, but it easier to have each module rune separately. But in the end modular monolith we have to one deployment).

meysamhadeli reviewed (approved):
Thanks for contribution :)
